File: _CsToDartTranspiler_/yarn_spinner/lib/src/yarn_spinner.markup/markup_parse_result.dart
/*

The MIT License (MIT)

Copyright (c) 2015-2017 Secret Lab Pty. Ltd. and Yarn Spinner contributors.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

*/

import 'package:yarn_spinner.framework/src/yarn_spinner.markup/markup_value_type.dart';
import 'package:yarn_spinner.framework/src/yarn_spinner.markup/tag_type.dart';


class MarkupParseResult {
  String Text;

  List<MarkupAttribute> Attributes;

  MarkupParseResult(String text, List<MarkupAttribute> attributes) {
    this.text = text;
    this.attributes = attributes;
  }

  bool tryGetAttributeWithName(String name, RefParam<MarkupAttribute> attribute) {
    for (var a in attributes) {
      if (a.name == name) {
        attribute.value = a;
        return true;
      }

    }

    attribute.value = default;
    return false;
  }

  String textForAttribute(MarkupAttribute attribute) {
    if (attribute.length == 0) {
      return String.empty;
    }


    if (text.length < attribute.position + attribute.length) {
      throw IndexOutOfRangeException("Attribute represents a range not representable by this text. Does this ${nameof(MarkupAttribute)} belong to this ${nameof(MarkupParseResult)}?");
    }


    return text.substring(attribute.position, attribute.length);
  }

  MarkupParseResult deleteRange(MarkupAttribute attributeToDelete) {
    var newAttributes = List<MarkupAttribute>();

    // Address the trivial case: if the attribute has a zero
    // length, just create a new markup that doesn't include it.
    // The plain text is left unmodified, because this attribute
    // didn't apply to any text.
    if (attributeToDelete.length == 0) {
      for (var a in attributes) {
        if (!a.equals(attributeToDelete)) {
          newAttributes.add(a);
        }

      }

      return MarkupParseResult(text, newAttributes);
    }


    var deletionStart = attributeToDelete.position;
    var deletionEnd = attributeToDelete.position + attributeToDelete.length;

    var editedSubstring = text.remove(attributeToDelete.position, attributeToDelete.length);

    for (var existingAttribute in attributes) {
      var start = existingAttribute.position;
      var end = existingAttribute.position + existingAttribute.length;

      if (existingAttribute.equals(attributeToDelete)) {
        // This is the attribute we're deleting. Don't include
        // it.
        continue
      }


      var editedAttribute = existingAttribute;

      if (start <= deletionStart) {
        // The attribute starts before start point of the item
        // we're deleting.
        if (end <= deletionStart) {
        }
        else if (end <= deletionEnd) {
          // This attribute starts before the item we're
          // deleting, and ends inside it. The Position
          // doesn't need to change, but its Length is
          // trimmed so that it ends where the deleted
          // attribute begins.
          editedAttribute.length = deletionStart - start;

          if (existingAttribute.length > 0 && editedAttribute.length <= 0) {
            // The attribute's length has been reduced to
            // zero. All of the contents it previous had
            // have been removed, so we will remove the
            // attribute itself.
            continue
          }

        }
        else {
          // This attribute starts before the item we're
          // deleting, and ends after it. Its length is
          // edited to remove the length of the item we're
          // deleting.
          editedAttribute.length -= attributeToDelete.length;
        }
      }
      else if (start >= deletionEnd) {
        // The item begins after the item we're deleting. Its
        // length isn't changing. We just need to offset its
        // start position.
        editedAttribute.position = start - attributeToDelete.length;
      }
      else if (start >= deletionStart && end <= deletionEnd) {
        // The item is entirely within the item we're deleting.
        // It will be deleted too - we'll skip including it in
        // the updated attributes list.
        continue
      }
      else if (start >= deletionStart && end > deletionEnd) {
        // The item starts within the item we're deleting, and
        // ends outside it. We'll adjust the start point so
        // that it begins at the point where this item and the
        // item we're deleting stop overlapping.
        var overlapLength = deletionEnd - start;
        var newStart = deletionStart;
        var newLength = existingAttribute.length - overlapLength;

        editedAttribute.position = newStart;
        editedAttribute.length = newLength;
      }


      newAttributes.add(editedAttribute);
    }

    return MarkupParseResult(editedSubstring, newAttributes);
  }
}

class MarkupAttribute {
  MarkupAttribute(int position, int sourcePosition, int length, String name, Iterable<MarkupProperty> properties) {
    this.position = position;
    this.sourcePosition = sourcePosition;
    this.length = length;
    this.name = name;

    var props = Map<String, MarkupValue>();

    for (var prop in properties) {
      props.add(prop.name, prop.value);
    }

    this.properties = props;
  }

  MarkupAttribute(MarkupAttributeMarker openingMarker, int length) {
  }

  int position = 0;

  int length = 0;

  String name = null;

  Map<String, MarkupValue> properties;

  int sourcePosition = 0;

  /// <inheritdoc
  String toString() {
    var sb = StringBuilder();
    sb.append("[${name}] - ${position}-${position + length} (${length}");

    if (properties?.count > 0) {
      sb.append(", ${properties.count} properties)");
    }


    sb.append(")");

    return sb.toString();
  }
}

class MarkupProperty {
  MarkupProperty(String name, MarkupValue value) {
    this.name = name;
    this.value = value;
  }

  String name = null;

  MarkupValue value = new MarkupValue();
}

class MarkupValue {
  /// Gets the integer value of this property.
  int integerValue = 0;

  /// Gets the float value of this property.
  double floatValue = 0.0;

  /// Gets the string value of this property.
  String stringValue = null;

  // Disable style warning "Summary should begin "Gets a value
  // indicating..." for this property, because that's not what this
  // bool property represents
  /// Gets the bool value of this property.
  bool boolValue = false;

  MarkupValueType type = MarkupValueType.Bool;

  /// <inheritdoc
  String toString() {
    switch (this.Type) {
      case MarkupValueType.integer: {
        return integerValue.toString();
      }
      case MarkupValueType.float: {
        return floatValue.toString();
      }
      case MarkupValueType.String: {
        return stringValue;
      }
      case MarkupValueType.bool: {
        return boolValue.toString();
      }
      default: {
        throw InvalidOperationException("Invalid markup value type ${type}");
      }
    }
  }
}

class MarkupAttributeMarker {
  MarkupAttributeMarker(String name, int position, int sourcePosition, List<MarkupProperty> properties, TagType type) {
    this.name = name;
    this.position = position;
    this.sourcePosition = sourcePosition;
    this.properties = properties;
    this.type = type;
  }

  String name = null;

  int position = 0;

  List<MarkupProperty> properties;

  TagType type = TagType.SelfClosing;

  int sourcePosition = 0;

  bool tryGetProperty(String name, RefParam<MarkupValue> result) {
    for (var prop in properties) {
      if (prop.name.equals(name)) {
        result.value = prop.value;
        return true;
      }

    }

    result.value = default;
    return false;
  }
}

File: YarnSpinner/YarnSpinner.Markup/MarkupParseResult.cs
namespace Yarn.Markup
{
    using System.Collections.Generic;

    /// <summary>
    /// The result of parsing a line of marked-up text.
    /// </summary>
    /// <remarks>
    /// You do not create instances of this struct yourself. It is created
    /// by objects that can parse markup, such as <see cref="Dialogue"/>.
    /// </remarks>
    /// <seealso cref="Dialogue.ParseMarkup(string)"/>
    public struct MarkupParseResult
    {
        /// <summary>
        /// The original text, with all parsed markers removed.
        /// </summary>
        public string Text;

        /// <summary>
        /// The list of <see cref="MarkupAttribute"/>s in this parse
        /// result.
        /// </summary>
        public List<MarkupAttribute> Attributes;

        /// <summary>
        /// Initializes a new instance of the <see cref="MarkupParseResult"/> struct.
        /// </summary>
        /// <param name="text">The plain text.</param>
        /// <param name="attributes">The list of attributes.</param>
        internal MarkupParseResult(string text, List<MarkupAttribute> attributes)
        {
            this.Text = text;
            this.Attributes = attributes;
        }

        /// <summary>
        /// Gets the first attribute with the specified name, if present.
        /// </summary>
        /// <param name="name">The name of the attribute to get.</param>
        /// <param name="attribute">When this method returns, contains the
        /// attribute with the specified name, if the attribute is found;
        /// otherwise, the default <see cref="MarkupAttribute"/>. This
        /// parameter is passed uninitialized.</param>
        /// <returns><see langword="true"/> if the <see
        /// cref="MarkupParseResult"/> contains an attribute with the
        /// specified name; otherwise, <see langword="false"/>.</returns>
        public bool TryGetAttributeWithName(string name, out MarkupAttribute attribute)
        {
            foreach (var a in this.Attributes)
            {
                if (a.Name == name)
                {
                    attribute = a;
                    return true;
                }
            }

            attribute = default;
            return false;
        }

        /// <summary>
        /// Returns the substring of <see cref="Text"/> covered by
        /// <paramref name="attribute"/> Position and Length properties.
        /// </summary>
        /// <remarks>
        /// <para>
        /// If the attribute's <see cref="MarkupAttribute.Length"/>
        /// property is zero, this method returns the empty string.
        /// </para>
        /// <para>
        /// This method does not check to see if <paramref
        /// name="attribute"/> is an attribute belonging to this
        /// MarkupParseResult. As a result, if you pass an attribute that
        /// doesn't belong, it may describe a range of text that does not
        /// appear in <see cref="Text"/>. If this occurs, an <see
        /// cref="System.IndexOutOfRangeException"/> will be thrown.
        /// </para>
        /// </remarks>
        /// <param name="attribute">The attribute to get the text
        /// for.</param>
        /// <returns>The text contained within the attribute.</returns>
        /// <throws cref="System.IndexOutOfRangeException">Thrown when
        /// attribute's <see cref="MarkupAttribute.Position"/> and <see
        /// cref="MarkupAttribute.Length"/> properties describe a range of
        /// text outside the maximum range of <see cref="Text"/>.</throws>
        public string TextForAttribute(MarkupAttribute attribute)
        {
            if (attribute.Length == 0)
            {
                return string.Empty;
            }

            if (this.Text.Length < attribute.Position + attribute.Length)
            {
                throw new System.IndexOutOfRangeException($"Attribute represents a range not representable by this text. Does this {nameof(MarkupAttribute)} belong to this {nameof(MarkupParseResult)}?");
            }

            return this.Text.Substring(attribute.Position, attribute.Length);
        }

        /// <summary>
        /// Deletes an attribute from this markup.
        /// </summary>
        /// <remarks>
        /// This method deletes the range of text covered by <paramref
        /// name="attributeToDelete"/>, and updates the other attributes in this
        /// markup as follows:
        ///
        /// <list type="bullet">
        /// <item>
        /// Attributes that start and end before the deleted attribute are
        /// unmodified.
        /// </item>
        ///
        /// <item>
        /// Attributes that start before the deleted attribute and end inside it
        /// are truncated to remove the part overlapping the deleted attribute.
        /// </item>
        ///
        /// <item>
        /// Attributes that have the same position and length as the deleted
        /// attribute are deleted, if they apply to any text.
        /// </item>
        ///
        /// <item>
        /// Attributes that start and end within the deleted attribute are
        /// deleted.
        /// </item>
        ///
        /// <item>
        /// Attributes that start within the deleted attribute, and end outside
        /// it, have their start truncated to remove the part overlapping the
        /// deleted attribute.
        /// </item>
        ///
        /// <item>
        /// Attributes that start after the deleted attribute have their start
        /// point adjusted to account for the deleted text.
        /// </item>
        /// </list>
        ///
        /// <para>
        /// This method does not modify the current object. A new <see
        /// cref="MarkupParseResult"/> is returned.
        /// </para>
        ///
        /// <para>
        /// If <paramref name="attributeToDelete"/> is not an attribute of this
        /// <see cref="MarkupParseResult"/>, the behaviour is undefined.
        /// </para>
        /// </remarks>
        /// <param name="attributeToDelete">The attribute to remove.</param>
        /// <returns>A new <see cref="MarkupParseResult"/> object, with the
        /// plain text modified and an updated collection of
        /// attributes.</returns>
        public MarkupParseResult DeleteRange(MarkupAttribute attributeToDelete)
        {
            var newAttributes = new List<MarkupAttribute>();

            // Address the trivial case: if the attribute has a zero
            // length, just create a new markup that doesn't include it.
            // The plain text is left unmodified, because this attribute
            // didn't apply to any text.
            if (attributeToDelete.Length == 0)
            {
                foreach (var a in this.Attributes)
                {
                    if (!a.Equals(attributeToDelete))
                    {
                        newAttributes.Add(a);
                    }
                }

                return new MarkupParseResult(this.Text, newAttributes);
            }

            var deletionStart = attributeToDelete.Position;
            var deletionEnd = attributeToDelete.Position + attributeToDelete.Length;

            var editedSubstring = this.Text.Remove(attributeToDelete.Position, attributeToDelete.Length);

            foreach (var existingAttribute in this.Attributes)
            {
                var start = existingAttribute.Position;
                var end = existingAttribute.Position + existingAttribute.Length;

                if (existingAttribute.Equals(attributeToDelete))
                {
                    // This is the attribute we're deleting. Don't include
                    // it.
                    continue;
                }

                var editedAttribute = existingAttribute;

                if (start <= deletionStart)
                {
                    // The attribute starts before start point of the item
                    // we're deleting.
                    if (end <= deletionStart)
                    {
                        // This attribute is entirely before the item we're
                        // deleting, and will be unmodified.
                    }
                    else if (end <= deletionEnd)
                    {
                        // This attribute starts before the item we're
                        // deleting, and ends inside it. The Position
                        // doesn't need to change, but its Length is
                        // trimmed so that it ends where the deleted
                        // attribute begins.
                        editedAttribute.Length = deletionStart - start;

                        if (existingAttribute.Length > 0 && editedAttribute.Length <= 0)
                        {
                            // The attribute's length has been reduced to
                            // zero. All of the contents it previous had
                            // have been removed, so we will remove the
                            // attribute itself.
                            continue;
                        }
                    }
                    else
                    {
                        // This attribute starts before the item we're
                        // deleting, and ends after it. Its length is
                        // edited to remove the length of the item we're
                        // deleting.
                        editedAttribute.Length -= attributeToDelete.Length;
                    }
                }
                else if (start >= deletionEnd)
                {
                    // The item begins after the item we're deleting. Its
                    // length isn't changing. We just need to offset its
                    // start position.
                    editedAttribute.Position = start - attributeToDelete.Length;
                }
                else if (start >= deletionStart && end <= deletionEnd)
                {
                    // The item is entirely within the item we're deleting.
                    // It will be deleted too - we'll skip including it in
                    // the updated attributes list.
                    continue;
                }
                else if (start >= deletionStart && end > deletionEnd)
                {
                    // The item starts within the item we're deleting, and
                    // ends outside it. We'll adjust the start point so
                    // that it begins at the point where this item and the
                    // item we're deleting stop overlapping.
                    var overlapLength = deletionEnd - start;
                    var newStart = deletionStart;
                    var newLength = existingAttribute.Length - overlapLength;

                    editedAttribute.Position = newStart;
                    editedAttribute.Length = newLength;
                }

                newAttributes.Add(editedAttribute);
            }

            return new MarkupParseResult(editedSubstring, newAttributes);
        }
    }
}
